src/_pytest/python.py

```diff
@@ -188,14 +188,21 @@ def pytest_pycollect_makemodule(path, parent):
     return Module(path, parent)
 
 
+def _safe_isclass(obj):
+    try:
+        return isclass(obj)
+    except Exception:
+        return False
```

Comment:
should this live in `pytest._compat`, that way other uses of `isclass` get the same benefits?

seems to be also used in these places:

```console
$ git grep compat.*isclass
src/_pytest/fixtures.py:from _pytest.compat import isclass
src/_pytest/python.py:from _pytest.compat import isclass
src/_pytest/python_api.py:from _pytest.compat import isclass
```

Reply:
I've thought it could be moved when needed.
Currently there are several places that handle things safely already - i.e. with the test case for django.conf.settings I've seen it coming to the "raise" at least once before.
Therefore I've thought to keep it near to the single user for now.

Reply:
I.e. I do not think we should wrap `isclass/isinstance` unnecessarily.

Reply:
:man_shrugging: it's triggerable in `raises` at least:

```python
import pytest

class CrappyClass(Exception):
    @property
    def __class__(self):
        assert False, 'nuuuu'


with pytest.raises(CrappyClass()):
    pass
```

```
Traceback (most recent call last):
  File "t.py", line 9, in <module>
    with pytest.raises(CrappyClass()):
  File "/tmp/pytest/venv/lib/python3.6/site-packages/_pytest/python_api.py", line 654, in raises
    for exc in filterfalse(isclass, always_iterable(expected_exception, BASE_TYPE)):
  File "/tmp/pytest/venv/lib/python3.6/site-packages/more_itertools/more.py", line 1305, in always_iterable
    if (base_type is not None) and isinstance(obj, base_type):
  File "t.py", line 6, in __class__
    assert False, 'nuuuu'
AssertionError: nuuuu
```

Reply:
Hmm, might be good to see this for `raises` after all?!  Not sure.
It's not as bad as with collection at least.

Reply:
I'll move it to compat and use it for your test case, too.

Reply:
Well, it fails there through `always_iterable` already, which does:

> if (base_type is not None) and isinstance(obj, base_type):

Fixing this to use `for exc in filterfalse(safe_isclass, always_iterable(expected_exception, base_type=None)):` (`base_type=None`) then makes it fail with:

> TypeError: exceptions must be old-style classes or derived from BaseException, not <class 'test_compat.test_safe_getclass.<locals>.CrappyClass'>

(beause it returns `False`)

Let's stick to "Arbitrary exceptions should not be silenced." here at least.